### PR TITLE
added support for identity attribute on container groups

### LINKED ISSUE #3241: Container Instance: Enable identity config
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

ACI has (preview) support for specifying identity of container groups upon creation. This behavior is very similar (identical?) to that for VMs, support for which already exists in this TF provider.

Note: I have a candidate branch that adds the feature for ACI, including tests and doc changes. Can share or push as desired (and permitted).

### New or Affected Resource(s)

azurerm_container_group

### Potential Terraform Configuration

```hcl
resource "azurerm_container_group" "my-aci-config" {
  name                = "some-name"
  .
  .
  .
  identity {
    type = "SystemAssigned"
  }
}
```

### References

* https://docs.microsoft.com/en-us/azure/container-instances/container-instances-managed-identity

* Similar to this for Logic Apps #2924

### PR DESCRIPTION
**Pull request by jplane**
Addresses this enhancement: https://github.com/terraform-providers/terraform-provider-azurerm/issues/3241

Fixes #3241

### CODE REVIEW

File: website/docs/r/container_group.html.markdown [hunk 1]
```diff
@@ -132,6 +134,16 @@ The following arguments are supported:
 
 ---
 
+An `identity` block supports the following:
+
+* `type` - (Required) The Managed Service Identity Type of this container group. Possible values are `SystemAssigned` (where Azure will generate a Service Principal for you), `UserAssigned` where you can specify the Service Principal IDs in the `identity_ids` field, and `SystemAssigned, UserAssigned` which assigns both a system managed identity as well as the specified user assigned identities.
```
Comment (tombuildsstuff) [inline website/docs/r/container_group.html.markdown]: (as above) can we confirm what this third value (`SystemAssigned, UserAssigned`) should be?

File: azurerm/resource_arm_container_group_test.go [hunk 2]
```diff
@@ -364,6 +456,121 @@ func TestAccAzureRMContainerGroup_windowsComplete(t *testing.T) {
...
+  container {
+    name   = "hw"
+    image  = "microsoft/aci-helloworld:latest"
+    cpu    = "0.5"
+    memory = "0.5"
+    port   = 80
+  }
+
+  identity {
+    type         = "SystemAssigned, UserAssigned"
```
[the remark this answers is not in the document]
Comment (jplane) [reply azurerm/resource_arm_container_group_test.go]: Good catch... the validation function is incorrect. Fixing and pushing now.

File: website/docs/r/container_group.html.markdown (hunk 1 above)
Comment (jplane) [reply website/docs/r/container_group.html.markdown]: 'SystemAssigned, UserAssigned' is correct... the validation function is wrong. Correcting now. See example in the Azure docs here:

https://docs.microsoft.com/en-us/azure/container-instances/container-instances-managed-identity#enable-managed-identity-using-yaml-file

File: azurerm/resource_arm_container_group.go [hunk 3]
```diff
@@ -709,6 +747,28 @@ func expandContainerEnvironmentVariables(input interface{}, secure bool) *[]cont
 	return &output
 }
 
+func expandContainerGroupIdentity(d *schema.ResourceData) *containerinstance.ContainerGroupIdentity {
+	v := d.Get("identity")
+	identities := v.([]interface{})
+	identity := identities[0].(map[string]interface{})
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_container_group.go]: if users don't have an identity block specified there's a crash here:

```
Test ended in panic.

------- Stdout: -------
=== RUN   TestAccAzureRMContainerGroup_imageRegistryCredentials
=== PAUSE TestAccAzureRMContainerGroup_imageRegistryCredentials
=== CONT  TestAccAzureRMContainerGroup_imageRegistryCredentials

------- Stderr: -------
panic: runtime error: index out of range

goroutine 361 [running]:
github.com/terraform-providers/terraform-provider-azurerm/azurerm.expandContainerGroupIdentity(0xc000362af0, 0xc0005b1680)
	/opt/teamcity-agent/work/458e5e4800bd94f6/src/github.com/terraform-providers/terraform-provider-azurerm/azurerm/resource_arm_container_group.go:753 +0x3a7
github.com/terraform-providers/terraform-provider-azurerm/azurerm.resourceArmContainerGroupCreate(0xc000362af0, 0x318f940, 0xc000cb0000, 0xc000362af0, 0x0)
	/opt/teamcity-agent/work/458e5e4800bd94f6/src/github.com/terraform-providers/terraform-provider-azurerm/azurerm/resource_arm_container_group.go:452 +0x483
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/helper/schema.(*Resource).Apply(0xc00052bb20, 0xc0000dbb80, 0xc000265000, 0x318f940, 0xc000cb0000, 0xc0008c8401, 0x3c, 0x0)
	/opt/teamcity-agent/work/458e5e4800bd94f6/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/helper/schema/resource.go:225 +0x351
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/helper/schema.(*Provider).Apply(0xc000579260, 0xc0009db950, 0xc0000dbb80, 0xc000265000, 0x1, 0x4813d3, 0x0)
	/opt/teamcity-agent/work/458e5e4800bd94f6/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/helper/schema/provider.go:283 +0x9c
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform.(*EvalApply).Eval(0xc000a941c0, 0x384a8a0, 0xc0009ad520, 0x2, 0x2, 0x3251849, 0x4)
	/opt/teamcity-agent/work/458e5e4800bd94f6/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform/eval_apply.go:57 +0x226
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform.EvalRaw(0x38029e0, 0xc000a941c0, 0x384a8a0, 0xc0009ad520, 0x0, 0x0, 0x0, 0x0)
	/opt/teamcity-agent/work/458e5e4800bd94f6/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform/eval.go:53 +0x156
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform.(*EvalSequence).Eval(0xc00074cf60, 0x384a8a0, 0xc0009ad520, 0x2, 0x2, 0x3251849, 0x4)
	/opt/teamcity-agent/work/458e5e4800bd94f6/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform/eval_sequence.go:14 +0x9c
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform.EvalRaw(0x3802f20, 0xc00074cf60, 0x384a8a0, 0xc0009ad520, 0x2e319e0, 0x627d3a2, 0x2d9d540, 0xc0006eda90)
	/opt/teamcity-agent/work/458e5e4800bd94f6/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform/eval.go:53 +0x156
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform.Eval(0x3802f20, 0xc00074cf60, 0x384a8a0, 0xc0009ad520, 0xc00074cf60, 0x3802f20, 0xc00074cf60, 0x34dc010)
	/opt/teamcity-agent/work/458e5e4800bd94f6/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform/eval.go:34 +0x4d
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform.(*Graph).walk.func1(0x3105000, 0xc000689450, 0x0, 0x0)
	/opt/teamcity-agent/work/458e5e4800bd94f6/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/terraform/graph.go:126 +0xc45
github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/dag.(*Walker).walkVertex(0xc0003927e0, 0x3105000, 0xc000689450, 0xc000ca41c0)
	/opt/teamcity-agent/work/458e5e4800bd94f6/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/dag/walk.go:387 +0x367
created by github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/dag.(*Walker).Update
	/opt/teamcity-agent/work/458e5e4800bd94f6/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/dag/walk.go:310 +0x986
```

can we make this:

```suggestion
	if len(identities) == 0 {
	  return nil
	}
	identity := identities[0].(map[string]interface{})
```

**Comment by tombuildsstuff** [comment]
@jplane unrelated to this PR - but it appears that the Windows Server docker images have been relocated from the Docker Hub to Microsoft's Container Registry; as such I'm going to push a commit to make those tests green, I hope you don't mind :)

**Comment by tombuildsstuff** [comment]
Ignoring the unrelated Windows test failures (mentioned above) the tests pass:

Screenshot 2019-04-18 at 12 06 53

**Comment by tombuildsstuff** [comment]
reverted the Windows container changes because this is a more involved/unrelated job

**Pull request merged**